variable/include/scipp/variable/variable_factory.h

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+#pragma once
+
+#include "scipp/variable/variable.h"
+
+namespace scipp::variable {
+
+/// Abstract base class for "variable makers", used by VariableFactory to
+/// dynamically create variables with given type.
+class SCIPP_VARIABLE_EXPORT AbstractVariableMaker {
+protected:
+  auto unreachable() const {
+    return std::logic_error("This code path should never be reached");
+  }
+
+public:
+  virtual ~AbstractVariableMaker() = default;
+  virtual bool is_buckets() const = 0;
+  virtual Variable
+  create(const DType elem_dtype, const Dimensions &dims, const bool variances,
+         const std::vector<VariableConstView> &parents) const = 0;
+  virtual DType elem_dtype(const VariableConstView &var) const = 0;
+  virtual bool hasVariances(const VariableConstView &var) const = 0;
+  virtual VariableConstView data(const VariableConstView &) const {
+    throw unreachable();
+  }
+  virtual VariableView data(const VariableView &) const { throw unreachable(); }
+  virtual core::element_array_view
```

Comment:
The reason is this: https://github.com/scipp/scipp/blob/2e05804a36d5efa85683fc82368e79acdf9c890e/variable/include/scipp/variable/variable_factory.h#L90-L97

We unfortunately need a special handling (lines 91 and 92) for the non-bucket case. I failed to find a nice way to do this using the virtual functions. The whole mechanism is a bit clunky, because we would need a templated virtual function (not possible), so the implementation is split up into two different virtual calls, to build up the typed views in the wrapping template. Can you see a better way?

Reply:
Not clear to me there is a better way.